// @neurosity/sdk/dist/cjs/utils/hapticEffects.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.buzz4_40 = exports.buzz3_60 = exports.buzz2_80 = exports.buzz1_100 = exports.longDoubleSharpTick3_60 = exports.longDoubleSharpTick2_80 = exports.longDoubleSharpTick1_100 = exports.longDoubleSharpClickMedium3_60 = exports.longDoubleSharpClickMedium2_80 = exports.longDoubleSharpClickMedium1_100 = exports.longDoubleSharpClickStrong4_30 = exports.longDoubleSharpClickStrong3_60 = exports.longDoubleSharpClickStrong2_80 = exports.longDoubleSharpClickStrong1_100 = exports.shortDoubleSharpTick3_60 = exports.shortDoubleSharpTick2_80 = exports.shortDoubleSharpTick1_100 = exports.shortDoubleClickMedium3_60 = exports.shortDoubleClickMedium2_80 = exports.shortDoubleClickMedium1_100 = exports.shortDoubleClickStrong4_30 = exports.shortDoubleClickStrong3_60 = exports.shortDoubleClickStrong2_80 = exports.shortDoubleClickStrong1_100 = exports.sharpTick3_60 = exports.sharpTick2_80 = exports.sharpTick1_100 = exports.mediumClick3_60 = exports.mediumClick2_80 = exports.mediumClick1_100 = exports.strongClick4_30 = exports.strongClick3_60 = exports.strongClick2_80 = exports.strongClick1_100 = exports.alert1000ms = exports.alert750ms = exports.strongBuzz100 = exports.softFuzz60 = exports.tripleClick100 = exports.doubleClick60 = exports.doubleClick100 = exports.softBump30 = exports.softBump60 = exports.softBump100 = exports.sharpClick30 = exports.sharpClick60 = exports.sharpClick100 = exports.strongClick30 = exports.strongClick60 = exports.strongClick100 = void 0;
exports.transitionRampDownLongSharp1_50_to_0 = exports.transitionRampDownShortSmooth2_50_to_0 = exports.transitionRampDownShortSmooth1_50_to_0 = exports.transitionRampDownMediumSmooth2_50_to_0 = exports.transitionRampDownMediumSmooth1_50_to_0 = exports.transitionRampDownLongSmooth2_50_to_0 = exports.transitionRampDownLongSmooth1_50_to_0 = exports.transitionRampUpShortSharp2_0_to_100 = exports.transitionRampUpShortSharp1_0_to_100 = exports.transitionRampUpMediumSharp2_0_to_100 = exports.transitionRampUpMediumSharp1_0_to_100 = exports.transitionRampUpLongSharp2_0_to_100 = exports.transitionRampUpLongSharp1_0_to_100 = exports.transitionRampUpShortSmooth2_0_to_100 = exports.transitionRampUpShortSmooth1_0_to_100 = exports.transitionRampUpMediumSmooth2_0_to_100 = exports.transitionRampUpMediumSmooth1_0_to_100 = exports.transitionRampUpLongSmooth2_0_to_100 = exports.transitionRampUpLongSmooth1_0_to_100 = exports.transitionRampDownShortSharp2_100_to_0 = exports.transitionRampDownShortSharp1_100_to_0 = exports.transitionRampDownLongMedium2_100_to_0 = exports.transitionRampDownLongMedium1_100_to_0 = exports.transitionRampDownLongSharp2_100_to_0 = exports.transitionRampDownLongSharp1_100_to_0 = exports.transitionRampDownShortSmooth2_100_to_0 = exports.transitionRampDownShortSmooth1_100_to_0 = exports.transitionRampDownMediumSmooth2_100_to_0 = exports.transitionRampDownMediumSmooth1_100_to_0 = exports.transitionRampDownLongSmooth2_100_to_0 = exports.transitionRampDownLongSmooth1_100_to_0 = exports.transistionHum6_10 = exports.transistionHum5_20 = exports.transistionHum4_40 = exports.transistionHum3_60 = exports.transistionHum2_80 = exports.transistionHum1_100 = exports.transistionClick6_10 = exports.transistionClick5_20 = exports.transistionClick4_40 = exports.transistionClick3_60 = exports.transistionClick2_80 = exports.transistionClick1_100 = exports.pulsingSharp2_60 = exports.pulsingSharp1_100 = exports.pulsingMedium2_60 = exports.pulsingMedium1_100 = exports.pulsingStong2_60 = exports.pulsingStong1_100 = exports.buzz5_20 = void 0;
exports.smoothHum5_10 = exports.smoothHum4_20 = exports.smoothHum3_30 = exports.smoothHum2_40 = exports.smoothHum1_50 = exports.longBuzzForProgrammaticStopping_100 = exports.transitionRampUpShortSharp2_0_to_50 = exports.transitionRampUpShortSharp1_0_to_50 = exports.transitionRampUpMediumSharp2_0_to_50 = exports.transitionRampUpMediumSharp1_0_to_50 = exports.transitionRampUpLongSharp2_0_to_50 = exports.transitionRampUpLongSharp1_0_to_50 = exports.transitionRampUpShortSmooth2_0_to_50 = exports.transitionRampUpShortSmooth1_0_to_50 = exports.transitionRampUpMediumSmooth2_0_to_50 = exports.transitionRampUpMediumSmooth1_0_to_50 = exports.transitionRampUpLongSmooth2_0_to_50 = exports.transitionRampUpLongSmooth1_0_to_50 = exports.transitionRampDownShortSharp2_50_to_0 = exports.transitionRampDownShortSharp1_50_to_0 = exports.transitionRampDownMediumSharp2_50_to_0 = exports.transitionRampDownMediumSharp1_50_to_0 = exports.transitionRampDownLongSharp2_50_to_0 = void 0;
// All 127ish codes here:
/// https://cdn-learn.adafruit.com/downloads/pdf/adafruit-drv2605-haptic-controller-breakout.pdf
exports.strongClick100 = "strongClick100";
exports.strongClick60 = "strongClick60";
exports.strongClick30 = "strongClick30";
exports.sharpClick100 = "sharpClick100";
exports.sharpClick60 = "sharpClick60";
exports.sharpClick30 = "sharpClick30";
exports.softBump100 = "softBump100";
exports.softBump60 = "softBump60";
exports.softBump30 = "softBump30";
exports.doubleClick100 = "doubleClick100";
exports.doubleClick60 = "doubleClick60";
exports.tripleClick100 = "tripleClick100";
exports.softFuzz60 = "softFuzz60";
exports.strongBuzz100 = "strongBuzz100";
exports.alert750ms = "alert750ms";
exports.alert1000ms = "alert1000ms";
exports.strongClick1_100 = "strongClick1_100";
exports.strongClick2_80 = "strongClick2_80";
exports.strongClick3_60 = "strongClick3_60";
exports.strongClick4_30 = "strongClick4_30";
exports.mediumClick1_100 = "mediumClick1_100";
exports.mediumClick2_80 = "mediumClick2_80";
exports.mediumClick3_60 = "mediumClick3_60";
exports.sharpTick1_100 = "sharpTick1_100";
exports.sharpTick2_80 = "sharpTick2_80";
exports.sharpTick3_60 = "sharpTick3_60";
exports.shortDoubleClickStrong1_100 = "shortDoubleClickStrong1_100";
exports.shortDoubleClickStrong2_80 = "shortDoubleClickStrong2_80";
exports.shortDoubleClickStrong3_60 = "shortDoubleClickStrong3_60";
exports.shortDoubleClickStrong4_30 = "shortDoubleClickStrong4_30";
exports.shortDoubleClickMedium1_100 = "shortDoubleClickMedium1_100";
exports.shortDoubleClickMedium2_80 = "shortDoubleClickMedium2_80";
exports.shortDoubleClickMedium3_60 = "shortDoubleClickMedium3_60";
exports.shortDoubleSharpTick1_100 = "shortDoubleSharpTick1_100";
exports.shortDoubleSharpTick2_80 = "shortDoubleSharpTick2_80";
exports.shortDoubleSharpTick3_60 = "shortDoubleSharpTick3_60";
exports.longDoubleSharpClickStrong1_100 = "longDoubleSharpClickStrong1_100";
exports.longDoubleSharpClickStrong2_80 = "longDoubleSharpClickStrong2_80";
exports.longDoubleSharpClickStrong3_60 = "longDoubleSharpClickStrong3_60";
exports.longDoubleSharpClickStrong4_30 = "longDoubleSharpClickStrong4_30";
exports.longDoubleSharpClickMedium1_100 = "longDoubleSharpClickMedium1_100";
exports.longDoubleSharpClickMedium2_80 = "longDoubleSharpClickMedium2_80";
exports.longDoubleSharpClickMedium3_60 = "longDoubleSharpClickMedium3_60";
exports.longDoubleSharpTick1_100 = "longDoubleSharpTick1_100";
exports.longDoubleSharpTick2_80 = "longDoubleSharpTick2_80";
exports.longDoubleSharpTick3_60 = "longDoubleSharpTick3_60";
exports.buzz1_100 = "buzz1_100";
exports.buzz2_80 = "buzz2_80";
exports.buzz3_60 = "buzz3_60";
exports.buzz4_40 = "buzz4_40";
exports.buzz5_20 = "buzz5_20";
exports.pulsingStong1_100 = "pulsingStong1_100";
exports.pulsingStong2_60 = "pulsingStong2_60";
exports.pulsingMedium1_100 = "pulsingMedium1_100";
exports.pulsingMedium2_60 = "pulsingMedium2_60";
exports.pulsingSharp1_100 = "pulsingSharp1_100";
exports.pulsingSharp2_60 = "pulsingSharp2_60";
exports.transistionClick1_100 = "transistionClick1_100";
exports.transistionClick2_80 = "transistionClick2_80";
exports.transistionClick3_60 = "transistionClick3_60";
exports.transistionClick4_40 = "transistionClick4_40";
exports.transistionClick5_20 = "transistionClick5_20";
exports.transistionClick6_10 = "transistionClick6_10";
exports.transistionHum1_100 = "transistionHum1_100";
exports.transistionHum2_80 = "transistionHum2_80";
exports.transistionHum3_60 = "transistionHum3_60";
exports.transistionHum4_40 = "transistionHum4_40";
exports.transistionHum5_20 = "transistionHum5_20";
exports.transistionHum6_10 = "transistionHum6_10";
exports.transitionRampDownLongSmooth1_100_to_0 = "transitionRampDownLongSmooth1_100_to_0";
exports.transitionRampDownLongSmooth2_100_to_0 = "transitionRampDownLongSmooth2_100_to_0";
exports.transitionRampDownMediumSmooth1_100_to_0 = "transitionRampDownMediumSmooth1_100_to_0";
exports.transitionRampDownMediumSmooth2_100_to_0 = "transitionRampDownMediumSmooth2_100_to_0";
exports.transitionRampDownShortSmooth1_100_to_0 = "transitionRampDownShortSmooth1_100_to_0";
exports.transitionRampDownShortSmooth2_100_to_0 = "transitionRampDownShortSmooth2_100_to_0";
exports.transitionRampDownLongSharp1_100_to_0 = "transitionRampDownLongSharp1_100_to_0";
exports.transitionRampDownLongSharp2_100_to_0 = "transitionRampDownLongSharp2_100_to_0";
exports.transitionRampDownLongMedium1_100_to_0 = "transitionRampDownLongMedium1_100_to_0";
exports.transitionRampDownLongMedium2_100_to_0 = "transitionRampDownLongMedium2_100_to_0";
exports.transitionRampDownShortSharp1_100_to_0 = "transitionRampDownShortSharp1_100_to_0";
exports.transitionRampDownShortSharp2_100_to_0 = "transitionRampDownShortSharp2_100_to_0";
exports.transitionRampUpLongSmooth1_0_to_100 = "transitionRampUpLongSmooth1_0_to_100";
exports.transitionRampUpLongSmooth2_0_to_100 = "transitionRampUpLongSmooth2_0_to_100";
exports.transitionRampUpMediumSmooth1_0_to_100 = "transitionRampUpMediumSmooth1_0_to_100";
exports.transitionRampUpMediumSmooth2_0_to_100 = "transitionRampUpMediumSmooth2_0_to_100";
exports.transitionRampUpShortSmooth1_0_to_100 = "transitionRampUpShortSmooth1_0_to_100";
exports.transitionRampUpShortSmooth2_0_to_100 = "transitionRampUpShortSmooth2_0_to_100";
exports.transitionRampUpLongSharp1_0_to_100 = "transitionRampUpLongSharp1_0_to_100";
exports.transitionRampUpLongSharp2_0_to_100 = "transitionRampUpLongSharp2_0_to_100";
exports.transitionRampUpMediumSharp1_0_to_100 = "transitionRampUpMediumSharp1_0_to_100";
exports.transitionRampUpMediumSharp2_0_to_100 = "transitionRampUpMediumSharp2_0_to_100";
exports.transitionRampUpShortSharp1_0_to_100 = "transitionRampUpShortSharp1_0_to_100";
exports.transitionRampUpShortSharp2_0_to_100 = "transitionRampUpShortSharp2_0_to_100";
exports.transitionRampDownLongSmooth1_50_to_0 = "transitionRampDownLongSmooth1_50_to_0";
exports.transitionRampDownLongSmooth2_50_to_0 = "transitionRampDownLongSmooth2_50_to_0";
exports.transitionRampDownMediumSmooth1_50_to_0 = "transitionRampDownMediumSmooth1_50_to_0";
exports.transitionRampDownMediumSmooth2_50_to_0 = "transitionRampDownMediumSmooth2_50_to_0";
exports.transitionRampDownShortSmooth1_50_to_0 = "transitionRampDownShortSmooth1_50_to_0";
exports.transitionRampDownShortSmooth2_50_to_0 = "transitionRampDownShortSmooth2_50_to_0";
exports.transitionRampDownLongSharp1_50_to_0 = "transitionRampDownLongSharp1_50_to_0";
exports.transitionRampDownLongSharp2_50_to_0 = "transitionRampDownLongSharp2_50_to_0";
exports.transitionRampDownMediumSharp1_50_to_0 = "transitionRampDownMediumSharp1_50_to_0";
exports.transitionRampDownMediumSharp2_50_to_0 = "transitionRampDownMediumSharp2_50_to_0";
exports.transitionRampDownShortSharp1_50_to_0 = "transitionRampDownShortSharp1_50_to_0";
exports.transitionRampDownShortSharp2_50_to_0 = "transitionRampDownShortSharp2_50_to_0";
exports.transitionRampUpLongSmooth1_0_to_50 = "transitionRampUpLongSmooth1_0_to_50";
exports.transitionRampUpLongSmooth2_0_to_50 = "transitionRampUpLongSmooth2_0_to_50";
exports.transitionRampUpMediumSmooth1_0_to_50 = "transitionRampUpMediumSmooth1_0_to_50";
exports.transitionRampUpMediumSmooth2_0_to_50 = "transitionRampUpMediumSmooth2_0_to_50";
exports.transitionRampUpShortSmooth1_0_to_50 = "transitionRampUpShortSmooth1_0_to_50";
exports.transitionRampUpShortSmooth2_0_to_50 = "transitionRampUpShortSmooth2_0_to_50";
exports.transitionRampUpLongSharp1_0_to_50 = "transitionRampUpLongSharp1_0_to_50";
exports.transitionRampUpLongSharp2_0_to_50 = "transitionRampUpLongSharp2_0_to_50";
exports.transitionRampUpMediumSharp1_0_to_50 = "transitionRampUpMediumSharp1_0_to_50";
exports.transitionRampUpMediumSharp2_0_to_50 = "transitionRampUpMediumSharp2_0_to_50";
exports.transitionRampUpShortSharp1_0_to_50 = "transitionRampUpShortSharp1_0_to_50";
exports.transitionRampUpShortSharp2_0_to_50 = "transitionRampUpShortSharp2_0_to_50";
exports.longBuzzForProgrammaticStopping_100 = "longBuzzForProgrammaticStopping_100";
exports.smoothHum1_50 = "smoothHum1_50";
exports.smoothHum2_40 = "smoothHum2_40";
exports.smoothHum3_30 = "smoothHum3_30";
exports.smoothHum4_20 = "smoothHum4_20";
exports.smoothHum5_10 = "smoothHum5_10";
